Docker image for scalatra + separate docker-compose.yml for production

jtnystrom opened the pull request:
This adds a docker image for Scalatra and docker-compose-prod.yml.

The scalatra docker image builds the scalatra server using ant, then copies the outputs to a new minimal image using a multi-stage build. The final image runs Scalatra on Jetty using the ScalatraLauncher that we also use for development.
I had to make the following tweaks:

* Distinguish an externally and internally reachable URL to fusionauth (since users need to be redirected to the former, but connecting internally may look different)
* Added a new ant target `scalatra-jetty` that compiles the back-end with jetty 
* Added FUSIONAUTH_APP_URL to .envrc (must now be set to run Scalatra, even when running with e.g. ant)

When we actually deploy we may make further tweaks, such as pull images from dockerhub instead of building directly from the current directory, and set local paths etc.

Test procedure:
1. If fuseki is to be run using docker, start fuseki, e.g. `docker-compose up -d fuseki ` (uses the usual docker-compose.yml), otherwise start separately
2. In .envrc, set T_TS_URL to http://fuseki:3030/Toxygates/query (pointing to the docker-compose.yml service called fuseki)  and similar for T_TS_UPDATE_URL since localhost:3030 isn't accessible from inside docker. (This means that we can connect to Fuseki either as another docker service, or as a standalone process.) Similarly, configure the FUSIONAUTH_APP_URL to also go to the docker service rather than to localhost.
3.  `docker-compose -f docker-compose-prod.yml up`
4. Following the first build, if the scalatra dockerimage needs to be updated, it is also necessary to do `docker-compose  -f docker-compose-prod.yml build scalatra` and then restarting this service

jtnystrom committed:
Improve comments for KCDBRegistry

jtnystrom committed:
Run Scalatra from docker-compose with a Dockerfile, work in progress

jtnystrom committed:
Distinguish internal fusionauth URL from the externally reachable one, using the new variable FUSIONAUTH_INTERNAL_URL.
Don't use scala's return statement inside a try-section, since this is implemented using a throwable, causing weird exceptions.

jtnystrom committed:
Pass additional environment variables to the Scalatra service from docker-compose.yml

jtnystrom committed:
Create separate docker-compose-prod.yml and move scalatra service there

jtnystrom committed:
Remove fuseki definition from prod docker-compose.

jtnystrom committed:
Pass T_TS_PASS to scalatra in prod docker-compose

jtnystrom committed:
Merge branch 'johan/docker' into johan/docker2

jtnystrom commented on OTGTool/src/main/scala/t/global/KCDBRegistry.scala:
```diff
@@ -25,37 +25,38 @@ import scala.collection.mutable.Map
  * The KCDBRegistry tracks open Kyoto Cabinet databases.
- * 
+ *
+ * Users can request readers and writers. Both should be closed after use.
```
I updated the comments in this file since I was thinking through what will happen if e.g. tomcat and scalatra both have write access to the kyoto cabinet databases, and I believe file locking will ensure nothing bad happens

jtnystrom commented on ivy.xml:
```diff
@@ -79,5 +80,8 @@
 
   <dependency org="io.fusionauth" name="fusionauth-java-client" rev="1.42.0" />
 
+  <dependency org="org.eclipse.jetty" name="jetty-webapp" rev="9.4.6.v20170531" conf="deploy->default" />
```
When we run scalatra from ant, we currently get these APIs on our classpath from gwt-devmode.jar, but when running Scalatra without either this or Tomcat they need to be supplied manually

jtnystrom commented on Toxygates/src/panomicon/Authentication.scala:
```diff
@@ -60,11 +64,11 @@ class Authentication {
   def getJwtToken(cookies: Array[Cookie]): Either[(JWT, String), String] = {
-    try {
-      val tokenCookie = cookies.find(c => c.getName == "__Host-jwt").getOrElse(
-        return Right("JWT cookie __Host-jwt was not set")
```
Returning inside a try-block (my mistake) is generally considered bad (compiles to a throwable and can have unpredictable results)

jtnystrom commented on docker-compose-prod.yml:
```diff
@@ -0,0 +1,86 @@
+    environment:
+      JVM_ARGS: ${SCALATRA_JVM_ARGS}
+      PORT: 8888
+      T_TS_URL: http://server-address:3030/Toxygates/query
```
server-address is a placeholder value and should be changed when deploying, depending on how we deploy.
If fuseki is running in docker, this can of course be http://fuseki:3030/Toxygates/... or similar

jtnystrom committed:
Remove kcchunk: prefix from the T_DATA_DIR configuration parameter (originally intended to distinguish between different database schemes,
but for some time we have only supported this scheme).
Revert the split of the T_DATA_DIR and T_DATA_PATH parameters (only the former is now needed).
Adjust URLs in docker-compose-prod.yml for easier testing together with docker-compose.yml.

yujiberra commented:
@jtnystrom 
> 2. Set fuseki URLs in docker-compose-prod.yml for the scalatra servlet (whether for external fuseki or internal service). If fuseki is started from the usual docker-compose.yml first, then something like http://jena-fuseki:3030/... should work.

Does this mean that `docker-compose-prod.yml` needs to be directly edited? If so, is there a reason why environment variables weren't used instead?

Also, in either case, I think a description of any additional environment variables that need to be set for `docker-compose-prod.yml` would be helpful.

jtnystrom commented:
@yujiberra The reason why I didn't use environment variables is that you can use this either with a standalone fuseki (like we currently run on the server) or you can use fuseki as a service in docker. I wanted to support both ways of running it.
If T_TS_URL is e.g. http://jena-fuseki:3030/... (where jena-fuseki is a docker service) then I thought it made sense to edit manually since that docker service might be specified by name in the same file.

That was my reasoning at the time, but on second thought, this feels a little half baked. It's probably more consistent to just do it with environment variables.

jtnystrom committed:
Rename FUSIONAUTH_INTERNAL_URL to FUSIONAUTH_APP_URL since this environment variable already existed and had the same meaning.
Document these variables in .envrc.example.

jtnystrom committed:
Scalatra service: take T_TS_URL, T_TS_UPDATE_URL and FUSIONAUTH_APP_URL from environment variables rather than hardcode in compose file

jtnystrom commented:
There are no new environment variables here now, but I renamed FUSIONAUTH_INTERNAL_URL to FUSIONAUTH_APP_URL which already existed (for the fusionauth service). I documented it in .envrc.example.

yujiberra commented:
@jtnystrom 
Thanks for your work on this; as I mentioned already I'll also write here some of the points we discussed yesterday:

>There are no new environment variables here now, but I renamed FUSIONAUTH_INTERNAL_URL to FUSIONAUTH_APP_URL which already existed (for the fusionauth service). I documented it in .envrc.example.

This note above should be in the PR description. (Ah right I see you did this yesterday, thanks!)

`docker/scalatra/panomicon` (or maybe some subdirectories) should be added to .gitignore

Regarding this part of the build procedure:
> 1. Build the resources for the dockerimage using ant dockerimage

I think the more standard way to do this would be for the Docker image creation to do the build. This would have a number of benefits:
- no need to have a specific ant build target just for making a docker image
- automate more of the build process
- the build artifact won't depend on the local environment
- no need for a local Java environment to do a build (could take advantage of this in the future to allow someone to do local development without a Java environment. I know I originally said we wouldn't pursue this, but having thought about it a bit more I think it might be beneficial, and shouldn't be that much work to set up anyway)

Doing this would probably require having the Dockerfile for scalatra in the Toxygates folder rather than in /docker/scalatra, but I think it's standard for the Dockerfile for an application to be in the same folder as its source files.

That's how I've set things up in my basic Docker configuration for Angular, which I've made a draft PR for here: https://github.com/Toxygates/Panomicon/pull/22

Also, regarding how we deal with Fuseki in the production docker configuration file, I wasn't actually able to get it to work following your steps (i.e., running Fuseki off the regular docker-compose, and then having the Scalatra container in docker-compose-prod talk to it) so let's discuss that some more at some point, as well as the possibility of configuring things differently.

jtnystrom committed:
Add docker/scalatra/panomicon/{classes,lib} to .gitignore

jtnystrom committed:
Rename the "dockerimage" ant target to "scalatra-jetty" to indicate its purpose better.
Set the build destination for this to Toxygates/scalatra-jetty and update .gitignore accordingly.

jtnystrom committed:
New Dockerfile for scalatra-jetty, which builds the scalatra server in the docker environment using a multi-stage build.
Added a special ivy_cache.xml to help with pre-populating the ivy cache when the docker image gets built (slow step).

jtnystrom commented:
@yujiberra Thanks for your comments, everything makes sense. I changed the Dockerfile to function as you suggested (as far as I could). This makes the docker image a lot slower to build, but as we've discussed, it's not intended to support frequent builds, and it's not so bad when it's been built once and the first steps can be cached.

I ended up putting the Dockerfile in Toxygates/, since that seems semantically most correct, but since it also needs files from e.g. OTGTool and from the shared root folder to build (I may restructure this some day), the context it is expecting is the root folder. So you could build it manually using e.g.
`docker build -f Toxygates/Dockerfile .`

I think the test process should work, but let me know if you still have problems.

jtnystrom committed:
Some cleanups for build.xml

jtnystrom committed:
Rename the ivy configuration 'deploy' to 'jetty' since this is clearer

jtnystrom committed:
Bugfix for scalatra docker image - invoke /bin/sh instead of /bin/bash

jtnystrom committed:
Update .envrc.example and add a comment

yujiberra commented on .envrc.example:
```diff
@@ -24,7 +25,11 @@ export T_INTERMINE_URL=https://targetmine.mizuguchilab.org
+#External URL that the user can be redirected to
 export FUSIONAUTH_BASEURL="http://localhost:9011"
+#Internal URL for connecting to fusionauth.
+export FUSIONAUTH_APP_URL="http://fusionauth:9011"
```
This should point to localhost, rather than fusionauth, to be consistent with lines 13-14. A comment similar to line 12 would be a good idea too.

jtnystrom committed:
Update .envrc.example to work out of the box with docker-compose.yml (FUSIONAUTH_APP_URL)

jtnystrom committed:
Base the scalatra Dockerfile on bitnami/minideb instead of alpine, since the latter contains musl instead of libc,
and this confuses jne-kyotocabinet when it tries to find the right native library

jtnystrom commented:
I changed the base image to minideb (a minimal debian based image) since the one I used was based on alpine, which uses musl instead of libc. This confused jne-kyotocabinet, and I should have tested more thoroughly and caught it earlier.

yujiberra reviewed (approved):
Thanks again for this work!
Not sure why I didn't end up approving since you made changes addressing all my comments, but in any case this looks reasonable (as far as I can judge based on my current memory of the application) and you can always make further fixes later on if necessary so I'm going to approve this.
